darts/explainability/shap_explainer.py

```diff
@@ -0,0 +1,511 @@
+        self.target_dim = self.model.input_dim["target"]
+
+        self.background_X = self._create_RegressionModel_shap_X(
+            background_series,
+            background_past_covariates,
+            background_future_covariates,
+            background_nb_samples,
+        )
+
+        self.target_dict = {c: i for i, c in enumerate(background_series.columns)}
+
+        if (self.target_dim > 1 or self.model.output_chunk_length > 1) and (
+            not self.multioutput
+        ):
+            self.explainers = {}
+            for i in range(self.model.output_chunk_length):
+                self.explainers[i] = {}
+                for j in range(self.target_dim):
+                    self.explainers[i][j] = shap.Explainer(
+                        self.model.model.estimators_[i + j], self.background_X
+                    )
+                    # Special case of trees, where we don't want to depend on a background (faster in general)
+                    # Warning: Randomforest is quite slow.
+                    if isinstance(self.explainers[i][j], shap.explainers._tree.Tree):
+                        self.explainers[i][j] = shap.TreeExplainer(
+                            self.model.model.estimators_[i + j]
+                        )
+        else:
+            self.explainers = shap.Explainer(self.model.model, self.background_X)
+            if isinstance(self.explainers, shap.explainers._tree.Tree):
+                self.explainers = shap.TreeExplainer(self.model.model)
+
+        self.cache_explainers = None
+        self.cache_foreground_series = None
+        self.cache_foreground_past_covariates = None
+        self.cache_foreground_future_covariates = None
+        self.cache_foreground_X = None
+        self.foreground_changed = True
+
+    def shap_values(
+        self,
+        foreground_series: Union[TimeSeries, Sequence[TimeSeries]],
+        foreground_past_covariates: Optional[Union[TimeSeries, Sequence[TimeSeries]]],
+        foreground_future_covariates: Optional[Union[TimeSeries, Sequence[TimeSeries]]],
+        horizon: Optional[int] = None,
+        target_name: Optional[str] = None,
+    ):
+        "-> shap._explanation.Explanation"
+
+        if not all(
+            foreground_series.time_index
+            == foreground_past_covariates.time_index
+            == foreground_future_covariates.time_index
+        ):
+            logger.warning(
+                "The series and covariates don't share the same time index. We will take the time index common to all."
+            )
+
+        (
+            foreground_series,
+            foreground_past_covariates,
+            foreground_future_covariates,
+        ) = retain_period_common_to_all(
+            [
+                foreground_series,
+                foreground_past_covariates,
+                foreground_future_covariates,
+            ]
+        )
+
+        # We don't recompute if the foreground is the same as the last computation
```

Comment:
Is it necessary to handle caching here? I would assume that it is users responsibility to avoid calling twice the function with the same inputs (but might be missing something)